File: src/enigma.R
# Enigma Machine — Rosetta Code Reference (R)
# Wehrmacht Enigma I: 3 rotors, Reflector B, plugboard.
# PeopleTec Inc. — Guinness World Record Attempt 2026

FWD <- c("EKMFLGDQVZNTOWYHXUSPAIBRCJ","AJDKSIRUXBLHWTMCQGZNPYFVOE","BDFHJLCPRTXVZNYEIWGAKMUSQO")
BWD <- c("UWYGADFPVZBECKMTHXSLRINQOJ","AJPCZWRLFBDKOTYUQGENHXMIVS","TAGBPCSDQEUFVNZHYIXJWLRKOM")
NOTCH <- c(81,69,86)  # Q, E, V
REFLECTOR <- "YRUHQSLDPXNGOKMIEBFZCWVJAT"

ci <- function(ch) utf8ToInt(ch) - 65L
cc <- function(i)  intToUtf8(i + 65L)

new_enigma <- function(rotors, key, plugboard=NULL) {
  e <- list(id=rotors+1L, off=ci(strsplit(key,"")[[1]]),
            plug=0:25)
  if (!is.null(plugboard)) {
    for (p in plugboard) {
      ab <- ci(strsplit(p,"")[[1]])
      e$plug[ab[1]+1] <- ab[2]; e$plug[ab[2]+1] <- ab[1]
    }
  }
  e
}

fwd_pass <- function(e, r, i) {
  c <- (i + e$off[r]) %% 26
  ch <- substr(FWD[e$id[r]], c+1, c+1)
  (ci(ch) - e$off[r] + 26) %% 26
}

bwd_pass <- function(e, r, i) {
  c <- (i + e$off[r]) %% 26
  ch <- substr(BWD[e$id[r]], c+1, c+1)
  (ci(ch) - e$off[r] + 26) %% 26
}

step_e <- function(e) {
  if (e$off[2] + 65 == NOTCH[e$id[2]]) {
    e$off[2] <- (e$off[2]+1) %% 26; e$off[1] <- (e$off[1]+1) %% 26
  } else if (e$off[3] + 65 == NOTCH[e$id[3]]) {
    e$off[2] <- (e$off[2]+1) %% 26
  }
  e$off[3] <- (e$off[3]+1) %% 26
  e
}

press_key <- function(e, ch) {
  e <- step_e(e)
  i <- ci(ch)
  i <- e$plug[i+1]
  i <- fwd_pass(e,3,i); i <- fwd_pass(e,2,i); i <- fwd_pass(e,1,i)
  i <- ci(substr(REFLECTOR, i+1, i+1))
  i <- bwd_pass(e,1,i); i <- bwd_pass(e,2,i); i <- bwd_pass(e,3,i)
  i <- e$plug[i+1]
  list(e=e, ch=cc(i))
}

encrypt <- function(e, text) {
  text <- gsub("[^A-Z]","",toupper(text))
  out <- ""
  for (j in seq_len(nchar(text))) {
    res <- press_key(e, substr(text,j,j))
    e <- res$e; out <- paste0(out, res$ch)
  }
  out
}

# Test vectors
tests <- list(
  list(c(0,1,2),"AAA",NULL,"AAAAA","BDZGO"),
  list(c(0,1,2),"AAA",NULL,"HELLOWORLD","ILBDAAMTAZ"),
  list(c(0,1,2),"AAA",NULL,"ATTACKATDAWN","BZHGNOCRRTCM"),
  list(c(0,1,2),"MCK",NULL,"HELLOWORLD","DLTBBQVPQV"),
  list(c(2,0,1),"AAA",NULL,"HELLOWORLD","KZHDFQYHXT"),
  list(c(0,1,2),"AAA",c("AB","CD","EF"),"HELLOWORLD","IKACBBMTBF")
)

all_ok <- TRUE
for (t in seq_along(tests)) {
  tv <- tests[[t]]
  e <- new_enigma(tv[[1]], tv[[2]], tv[[3]])
  ct <- encrypt(e, tv[[4]])
  ok <- ct == tv[[5]]
  cat(sprintf("Test %d: %-20s -> %-15s [%s]\n", t, tv[[4]], ct, ifelse(ok,"PASS","FAIL")))
  if (!ok) all_ok <- FALSE
}
cat(ifelse(all_ok, "\nALL 6 TESTS PASSED\n", "\nSOME TESTS FAILED\n"))

File: src/enigma.go
// Enigma Machine — Rosetta Code Reference (Go)
// Wehrmacht Enigma I: 3 rotors, Reflector B, plugboard.
// PeopleTec Inc. — Guinness World Record Attempt 2026
package main

import (
	"fmt"
	"strings"
)

var (
	fwd  = [3]string{"EKMFLGDQVZNTOWYHXUSPAIBRCJ", "AJDKSIRUXBLHWTMCQGZNPYFVOE", "BDFHJLCPRTXVZNYEIWGAKMUSQO"}
	bwd  = [3]string{"UWYGADFPVZBECKMTHXSLRINQOJ", "AJPCZWRLFBDKOTYUQGENHXMIVS", "TAGBPCSDQEUFVNZHYIXJWLRKOM"}
	ntch = [3]byte{'Q', 'E', 'V'}
	refl = "YRUHQSLDPXNGOKMIEBFZCWVJAT"
)

type Enigma struct {
	id, off [3]int
	plug    [26]int
}

func NewEnigma(rotors [3]int, key string, plugs []string) *Enigma {
	e := &Enigma{id: rotors}
	for i := 0; i < 3; i++ { e.off[i] = int(key[i] - 'A') }
	for i := 0; i < 26; i++ { e.plug[i] = i }
	for _, p := range plugs {
		a, b := int(p[0]-'A'), int(p[1]-'A')
		e.plug[a], e.plug[b] = b, a
	}
	return e
}

func (e *Enigma) fwdPass(r, i int) int {
	c := (i + e.off[r]) % 26
	return (int(fwd[e.id[r]][c]-'A') - e.off[r] + 26) % 26
}
func (e *Enigma) bwdPass(r, i int) int {
	c := (i + e.off[r]) % 26
	return (int(bwd[e.id[r]][c]-'A') - e.off[r] + 26) % 26
}

func (e *Enigma) step() {
	if byte(e.off[1])+'A' == ntch[e.id[1]] {
		e.off[1] = (e.off[1] + 1) % 26
		e.off[0] = (e.off[0] + 1) % 26
	} else if byte(e.off[2])+'A' == ntch[e.id[2]] {
		e.off[1] = (e.off[1] + 1) % 26
	}
	e.off[2] = (e.off[2] + 1) % 26
}

func (e *Enigma) pressKey(ch byte) byte {
	e.step()
	i := int(ch - 'A')
	i = e.plug[i]
	i = e.fwdPass(2, i); i = e.fwdPass(1, i); i = e.fwdPass(0, i)
	i = int(refl[i] - 'A')
	i = e.bwdPass(0, i); i = e.bwdPass(1, i); i = e.bwdPass(2, i)
	i = e.plug[i]
	return byte(i) + 'A'
}

func (e *Enigma) Encrypt(text string) string {
	var sb strings.Builder
	for _, c := range strings.ToUpper(text) {
		if c >= 'A' && c <= 'Z' { sb.WriteByte(e.pressKey(byte(c))) }
	}
	return sb.String()
}

func main() {
	type tv struct{ r [3]int; k string; p []string; pt, exp string }
	tests := []tv{
		{[3]int{0,1,2},"AAA",nil,"AAAAA","BDZGO"},
		{[3]int{0,1,2},"AAA",nil,"HELLOWORLD","ILBDAAMTAZ"},
		{[3]int{0,1,2},"AAA",nil,"ATTACKATDAWN","BZHGNOCRRTCM"},
		{[3]int{0,1,2},"MCK",nil,"HELLOWORLD","DLTBBQVPQV"},
		{[3]int{2,0,1},"AAA",nil,"HELLOWORLD","KZHDFQYHXT"},
		{[3]int{0,1,2},"AAA",[]string{"AB","CD","EF"},"HELLOWORLD","IKACBBMTBF"},
	}
	allOk := true
	for i, t := range tests {
		e := NewEnigma(t.r, t.k, t.p)
		ct := e.Encrypt(t.pt)
		ok := ct == t.exp
		status := "PASS"; if !ok { status = "FAIL"; allOk = false }
		fmt.Printf("Test %d: %-20s -> %-15s [%s]\n", i+1, t.pt, ct, status)
	}
	if allOk { fmt.Println("\nALL 6 TESTS PASSED") } else { fmt.Println("\nSOME TESTS FAILED") }
}
